### PR TITLE
fix: json.set recursive processing crash

### PR DESCRIPTION
**Pull request by vyavdoshenko**
Fixes: https://github.com/dragonflydb/dragonfly/issues/5038

1. What's not working:
Using JSON.SET with recursive descent paths ($..) on nested JSON objects could lead to hangs or crashes. This occurred because the mutation logic could process the same node multiple times if it was discovered via different paths during the traversal.

2. How this PR fixes the issue:
This PR adjusts the iteration logic. When processing elements of an object or array targeted by a path segment, the code now correctly handles iterator advancement, especially after an element might have been modified or erased by the mutation callback. This ensures that each element within the target object/array is considered for mutation exactly once during that step, preventing infinite loops or incorrect processing that previously led to the crash/hang with $.. paths.

### CODE REVIEW

**Comment by romange** [comment]
Please add explanation to the PR description. In general, for any non-trivial PRs, please state at least:
1. what's not working
2. how this PR fixes the problem.

File: src/server/json_family_test.cc [hunk 1]
```diff
@@ -3087,4 +3087,29 @@ TEST_F(JsonFamilyTest, DepthLimitExceeded) {
   ASSERT_THAT(resp, ErrArg("ERR failed to parse JSON"));
 }
 
+TEST_F(JsonFamilyTest, RecursiveDescentMutation) {
```
Comment (romange) [inline src/server/json_family_test.cc:3090]: Once you identified the problem you should test what is the minimum possible unit test that can help reproducing the problem. It's fine to add tests here but it's necessary to add tests to `src/core/json/jsonpath_test.cc` where you can reproduce it precisely and be able to iterate quickly.

File: src/core/json/detail/jsoncons_dfs.cc [hunk 2]
```diff
@@ -8,6 +8,8 @@
 
 #include "core/json/detail/jsoncons_dfs.h"
 
+#include <set>
```
Comment (romange) [inline src/core/json/detail/jsoncons_dfs.cc:11]: we use `#include <absl/container/flat_hash_set.h>` as drop in replacement

**Review by romange — COMMENTED** [review]
good work simplifying the fix!

File: src/server/json_family_test.cc (hunk 1 above)
Comment (vyavdoshenko) [reply src/server/json_family_test.cc]: Fixed

File: src/core/json/detail/jsoncons_dfs.h [hunk 3]
```diff
@@ -43,6 +43,16 @@ template <bool IsConst> class JsonconsDfsItem {
     return depth_state_.second;
   }
 
+  // Added getter for the object pointer
+  Ptr obj_ptr() const {
+    return depth_state_.first;
+  }
+
+  // Added getter for segment_step_
```
[the remark this answers is not in the document]
Comment (vyavdoshenko) [reply src/core/json/detail/jsoncons_dfs.h]: Fixed

File: src/core/json/detail/jsoncons_dfs.cc (hunk 2 above)
Comment (vyavdoshenko) [reply src/core/json/detail/jsoncons_dfs.cc]: Fixed

File: src/core/json/detail/jsoncons_dfs.cc [hunk 4]
```diff
@@ -103,14 +109,35 @@ Dfs Dfs::Mutate(absl::Span<const PathSegment> path, const MutateCallback& callba
         if (next_seg_id + 1 < path.size()) {
           stack.emplace_back(next, next_seg_id);
         } else {
-          dfs.MutateStep(path[next_seg_id], callback, next);
+          // Terminal step: collect node for mutation if not seen before
+          if (seen_nodes.find(next) == seen_nodes.end()) {
```
Comment (romange) [inline src/core/json/detail/jsoncons_dfs.cc:113]: I am curious actually. How come we can meet the same node twice? is `seen_nodes` needed at all?

Comment (vyavdoshenko) [reply src/core/json/detail/jsoncons_dfs.cc:113]: Yes, it's possible. I don't know how it can be done using JSON itself. But using modifications, we can get different paths to the same object. Also, it can be possible using a key and/or recursive replacement.

Comment (romange) [reply src/core/json/detail/jsoncons_dfs.cc:113]: Fair enough. Do we have a json path test that reproduces this edge case, i.e. would check fail if we add 
`DCHECK(seen_nodes.find(next) == seen_nodes.end())` ?

Comment (romange) [reply src/core/json/detail/jsoncons_dfs.cc:113]: Also why do we need both vector and a set for this? why can not we iterate over set below?

Comment (romange) [reply src/core/json/detail/jsoncons_dfs.cc:113]: Could you add a test that proves that `seen_nodes` are required?

File: src/core/json/detail/jsoncons_dfs.cc [hunk 5]
```diff
@@ -82,6 +84,10 @@ Dfs Dfs::Mutate(absl::Span<const PathSegment> path, const MutateCallback& callba
     return dfs;
   }
 
+  // Use vector to maintain order and set to check uniqueness
```
Comment (romange) [inline src/core/json/detail/jsoncons_dfs.cc:87]: I missed this comment

File: src/core/json/detail/jsoncons_dfs.cc (hunk 4 above)
Comment (vyavdoshenko) [reply src/core/json/detail/jsoncons_dfs.cc]: I tried to verify it using tests, but no matter how hard I tried, I couldn’t get it to work — and there’s an explanation for that.
Jsoncons doesn’t reuse the same object; instead, it creates a new one each time it’s needed.
So, my assumption was incorrect, and the set was unnecessary.

**Review by romange — APPROVED** [review]
